movebucket_test.go

```diff
@@ -0,0 +1,290 @@
+				}
+
+				return nil
+			}); err != nil {
+				t.Fatal(err)
+			}
+			db.MustCheck()
+
+			// act
+			if err := db.Update(func(tx *bbolt.Tx) error {
+				srcBucket := retrieveBucket(t, tx, tc.srcBucketPath...)
+				dstBucket := retrieveBucket(t, tx, tc.dstBucketPath...)
+
+				var bucketToMove *bbolt.Bucket
+				if srcBucket != nil {
+					bucketToMove = srcBucket.Bucket([]byte(tc.bucketToMove))
+				} else {
+					bucketToMove = tx.Bucket([]byte(tc.bucketToMove))
+				}
+
+				if bucketToMove != nil {
+					if wErr := dumpBucket([]byte(tc.bucketToMove), bucketToMove, dumpBucketBeforeMoving); wErr != nil {
+						t.Fatalf("error dumping bucket %v to file %v: %v", bucketToMove.String(), dumpBucketBeforeMoving, wErr)
+					}
+				}
+
+				mErr := tx.MoveBucket([]byte(tc.bucketToMove), srcBucket, dstBucket)
+				require.ErrorIs(t, mErr, tc.expActErr)
+
+				return nil
+			}); err != nil {
+				t.Fatal(err)
+			}
+			db.MustCheck()
+
+			// skip assertion if failure expected
+			if tc.expActErr != nil {
+				return
+			}
+
+			// assert
+			if err := db.View(func(tx *bbolt.Tx) error {
+				var movedBucket *bbolt.Bucket
+				srcBucket := retrieveBucket(t, tx, tc.srcBucketPath...)
+
+				if srcBucket != nil {
+					if movedBucket = srcBucket.Bucket([]byte(tc.bucketToMove)); movedBucket != nil {
+						t.Fatalf("expected childBucket %v to be moved from srcBucket %v", tc.bucketToMove, srcBucket)
+					}
+				} else {
+					if movedBucket = tx.Bucket([]byte(tc.bucketToMove)); movedBucket != nil {
+						t.Fatalf("expected childBucket %v to be moved from root bucket %v", tc.bucketToMove, "root bucket")
+					}
+				}
+
+				dstBucket := retrieveBucket(t, tx, tc.dstBucketPath...)
+				if dstBucket != nil {
+					if movedBucket = dstBucket.Bucket([]byte(tc.bucketToMove)); movedBucket == nil {
+						t.Fatalf("expected childBucket %v to be child of dstBucket %v", tc.bucketToMove, dstBucket)
+					}
+				} else {
+					if movedBucket = tx.Bucket([]byte(tc.bucketToMove)); movedBucket == nil {
+						t.Fatalf("expected childBucket %v to be child of dstBucket %v", tc.bucketToMove, "root bucket")
+					}
+				}
+
+				wErr := dumpBucket([]byte(tc.bucketToMove), movedBucket, dumpBucketAfterMoving)
+				if wErr != nil {
+					t.Fatalf("error dumping bucket %v to file %v", movedBucket.String(), dumpBucketAfterMoving)
+				}
+
+				beforeBucket := readBucketFromFile(t, dumpBucketBeforeMoving)
+				afterBucket := readBucketFromFile(t, dumpBucketAfterMoving)
+
+				if !bytes.Equal(beforeBucket, afterBucket) {
+					t.Fatalf("bucket's content before moving is different than after moving")
+				}
+
+				return nil
+			}); err != nil {
+				t.Fatal(err)
+			}
+			db.MustCheck()
+		})
+	}
+}
+
+func createBucketIfNotExist(t testing.TB, tx *bbolt.Tx, incompatibleKey bool, paths ...string) *bbolt.Bucket {
+	t.Helper()
+
+	var bk *bbolt.Bucket
+	var err error
+
+	idx := len(paths) - 1
+	for i, key := range paths {
+		if len(key) == 0 || (incompatibleKey && i == idx) {
```

Comment:
@ahrtr  this extra check is to avoid cluttering the `arrange` phase with logic 

- if the bucketKey should be normal K/V, then we skip creating the last bucket in the path
- otherwise we create the subBucket